File: electron/services/session-registry.ts
import { logger } from "../utils/logger";

interface SessionEntry {
  threadId: string;
  sessionId: string;
  createdAt: string;
}

/**
 * Maps PeriCode thread IDs to Claude Agent SDK session IDs.
 * Enables resume and fork operations across app restarts.
 */
class SessionRegistry {
  private sessions: Map<string, SessionEntry> = new Map();

  register(threadId: string, sessionId: string): void {
    this.sessions.set(threadId, {
      threadId,
      sessionId,
      createdAt: new Date().toISOString(),
    });
    logger.debug("session-registry", `Registered session ${sessionId} for thread ${threadId}`);
  }

  getSessionId(threadId: string): string | null {
    return this.sessions.get(threadId)?.sessionId ?? null;
  }

  getThreadId(sessionId: string): string | null {
    for (const entry of this.sessions.values()) {
      if (entry.sessionId === sessionId) {
        return entry.threadId;
      }
    }
    return null;
  }

  remove(threadId: string): void {
    this.sessions.delete(threadId);
  }

  hasSession(threadId: string): boolean {
    return this.sessions.has(threadId);
  }

  getAllSessions(): SessionEntry[] {
    return Array.from(this.sessions.values());
  }

  clear(): void {
    this.sessions.clear();
  }
}

export const sessionRegistry = new SessionRegistry();

File: src-tauri/src/services/session_registry.rs
use tracing::debug;

use crate::state::{AppState, SessionEntry};

/// Register a Claude CLI session for a given thread.
pub fn register(state: &AppState, thread_id: &str, session_id: &str) {
    let mut sessions = state.sessions.write();
    sessions.insert(
        thread_id.to_string(),
        SessionEntry {
            thread_id: thread_id.to_string(),
            session_id: session_id.to_string(),
            created_at: chrono::Utc::now().to_rfc3339(),
        },
    );
    debug!(
        target: "session-registry",
        "Registered session {} for thread {}", session_id, thread_id
    );
}

/// Get the CLI session ID for a given thread.
pub fn get_session_id(state: &AppState, thread_id: &str) -> Option<String> {
    let sessions = state.sessions.read();
    sessions.get(thread_id).map(|e| e.session_id.clone())
}

/// Look up which thread owns a given session ID.
pub fn get_thread_id(state: &AppState, session_id: &str) -> Option<String> {
    let sessions = state.sessions.read();
    for entry in sessions.values() {
        if entry.session_id == session_id {
            return Some(entry.thread_id.clone());
        }
    }
    None
}

/// Remove a session entry for the given thread.
pub fn remove(state: &AppState, thread_id: &str) {
    let mut sessions = state.sessions.write();
    sessions.remove(thread_id);
}

/// Check whether a session is registered for the given thread.
pub fn has_session(state: &AppState, thread_id: &str) -> bool {
    let sessions = state.sessions.read();
    sessions.contains_key(thread_id)
}

/// Return a snapshot of all registered sessions.
pub fn get_all_sessions(state: &AppState) -> Vec<SessionEntry> {
    let sessions = state.sessions.read();
    sessions.values().cloned().collect()
}

/// Clear all registered sessions.
pub fn clear(state: &AppState) {
    let mut sessions = state.sessions.write();
    sessions.clear();
}
